src/check_deps.rs

```diff
@@ -0,0 +1,97 @@
+use crate::utils::rename_deps;
+use std::collections::BTreeMap;
+use std::fs;
+use std::path::Path;
+use toml_edit::DocumentMut;
+pub fn check_dependencies(
+    cargo_toml_path: &Path,
+    crates_versions: &BTreeMap<String, String>,
+    overwrite: bool,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let _ = check_dependencies_impl(cargo_toml_path, crates_versions, overwrite)?;
+    log::info!(
+        "Checked for dependencies in {}. All Up-to-date!",
+        cargo_toml_path.display()
+    );
+
+    Ok(())
+}
+
+pub fn check_dependencies_impl(
+    cargo_toml_path: &Path,
+    crates_versions: &BTreeMap<String, String>,
+    overwrite: bool,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let local_toml_content = fs::read_to_string(cargo_toml_path)?;
+    let mut cargo_toml: DocumentMut = local_toml_content.parse()?;
+    let mut has_mismatch = false;
+
+    // Check if cargo workspace is defined
+    let deps = match cargo_toml.as_table_mut().get_mut("workspace") {
+        Some(toml_edit::Item::Table(table)) => table,
+        _ => cargo_toml.as_table_mut(),
+    };
+
+    for table in ["dependencies", "dev-dependencies", "build-dependencies"].iter() {
+        if let Some(toml_edit::Item::Table(dep_table)) = deps.get_mut(table) {
+            has_mismatch = check_table_dependencies(dep_table, crates_versions, overwrite)?;
+        }
+    }
+
+    if has_mismatch {
+        return Err("One or More Dependency version mismatch found".into());
+    }
+
+    Ok(())
+}
+
+fn check_table_dependencies(
+    dep_table: &mut toml_edit::Table,
+    crates_versions: &BTreeMap<String, String>,
+    overwrite: bool,
+) -> Result<bool, Box<dyn std::error::Error>> {
+    let mut local_toml_version: &str = "";
```

Comment:
To get rid of the warning since we are not reading the value until we assign it.

```suggestion
    let mut local_toml_version: &str;
```

Reply:
I fixed the warnings you mentioned, should be good now. If possible can you point out the code duplications? I took the liberty in changing the original code structure to prevent some code duplication and modularise it. But if there's still some duplication, I would love to finish it up before this is merged.

Reply:
This is probably the approach I would have taken. This way, we would only need to add a new `check_dependency` function where your checking logic would be placed.

NOTE: this is a diff against the main branch.

```diff
diff --git a/src/main.rs b/src/main.rs
index b0dd6b0..79cbaa3 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -49,6 +49,10 @@ struct Command {
     /// List available versions.
     #[clap(short, long)]
     list: bool,
+
+    /// Checks if deps have the same version as the specified Polkadot SDK version.
+    #[clap(short, long)]
+    check: bool,
 }
 
 #[tokio::main]
@@ -73,7 +77,12 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let crates_versions: BTreeMap<String, String> =
         get_version_mapping_with_fallback(DEFAULT_GIT_SERVER, &version).await?;
 
-    update_dependencies(&cargo_toml_path, &crates_versions, cmd.overwrite)?;
+    check_dependencies(
+        &cargo_toml_path,
+        &crates_versions,
+        cmd.overwrite,
+        !cmd.check,
+    )?;
 
     Ok(())
 }
@@ -94,12 +103,13 @@ fn validate_workspace_path(mut path: PathBuf) -> Result<PathBuf, Box<dyn std::er
     Ok(path)
 }
 
-fn update_dependencies(
+fn check_dependencies(
     cargo_toml_path: &Path,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) -> Result<(), Box<dyn std::error::Error>> {
-    let cargo_toml = update_dependencies_impl(cargo_toml_path, crates_versions, overwrite)?;
+    let cargo_toml = check_dependencies_impl(cargo_toml_path, crates_versions, overwrite, update)?;
 
     match cargo_toml {
         Some(new_content) => {
@@ -117,10 +127,11 @@ fn update_dependencies(
     Ok(())
 }
 
-fn update_dependencies_impl(
+fn check_dependencies_impl(
     cargo_toml_path: &Path,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) -> Result<Option<String>, Box<dyn std::error::Error>> {
     let cargo_toml_content = fs::read_to_string(cargo_toml_path)?;
     let mut cargo_toml: DocumentMut = cargo_toml_content.parse()?;
@@ -132,7 +143,7 @@ fn update_dependencies_impl(
 
     for table in ["dependencies", "dev-dependencies", "build-dependencies"].iter() {
         if let Some(toml_edit::Item::Table(dep_table)) = deps.get_mut(table) {
-            update_table_dependencies(dep_table, crates_versions, overwrite);
+            check_table_dependencies(dep_table, crates_versions, overwrite, update);
         }
     }
 
@@ -144,10 +155,11 @@ fn update_dependencies_impl(
     }
 }
 
-pub fn update_table_dependencies(
+pub fn check_table_dependencies(
     dep_table: &mut toml_edit::Table,
     crates_versions: &BTreeMap<String, String>,
     overwrite: bool,
+    update: bool,
 ) {
     for (dep_key, dep_value) in dep_table.iter_mut() {
         let dep_key_str = dep_key.get();
@@ -167,42 +179,55 @@ pub fn update_table_dependencies(
             continue;
         };
 
-        if let Some(table) = dep_value.as_table_like_mut() {
-            if !overwrite && table.get("path").is_some() {
-                continue;
-            }
+        if update {
+            update_dependency(dep_value, dep_key_str, crate_version, overwrite);
+        } else {
+            // TODO: check
+        }
+    }
+}
 
-            table.remove("git");
-            table.remove("rev");
-            table.remove("branch");
-            table.remove("tag");
-            table.remove("path");
+fn update_dependency(
+    dep_value: &mut toml_edit::Item,
+    dep_key_str: &str,
+    crate_version: &String,
+    overwrite: bool,
+) {
+    if let Some(table) = dep_value.as_table_like_mut() {
+        if !overwrite && table.get("path").is_some() {
+            return;
+        }
 
-            let mut new_table = toml_edit::InlineTable::default();
+        table.remove("git");
+        table.remove("rev");
+        table.remove("branch");
+        table.remove("tag");
+        table.remove("path");
 
-            // Directly create a `toml_edit::Value` for the version
-            new_table.get_or_insert(
-                "version",
-                toml_edit::value(crate_version.clone()).as_value().unwrap(),
-            );
+        let mut new_table = toml_edit::InlineTable::default();
 
-            for (key, value) in table.iter() {
-                // Ensure we're inserting `Value`s, not `Item`s
-                if key != "version" && value.is_value() {
-                    new_table.get_or_insert(key, value.as_value().unwrap().clone());
-                }
-            }
-            new_table.fmt();
+        // Directly create a `toml_edit::Value` for the version
+        new_table.get_or_insert(
+            "version",
+            toml_edit::value(crate_version.clone()).as_value().unwrap(),
+        );
 
-            // Replace the original table-like item with the new inline table
-            *dep_value = toml_edit::Item::Value(toml_edit::Value::InlineTable(new_table));
-        } else if dep_value.is_str() {
-            *dep_value = toml_edit::value(crate_version.clone());
-        } else {
-            log::error!("Unexpected dependency value type for {}", dep_key_str);
-            continue;
+        for (key, value) in table.iter() {
+            // Ensure we're inserting `Value`s, not `Item`s
+            if key != "version" && value.is_value() {
+                new_table.get_or_insert(key, value.as_value().unwrap().clone());
+            }
         }
+        new_table.fmt();
 
-        log::debug!("Setting {} to {}", dep_key_str, crate_version);
+        // Return the new inline table
+        *dep_value = toml_edit::Item::Value(toml_edit::Value::InlineTable(new_table))
+    } else if dep_value.is_str() {
+        *dep_value = toml_edit::value(crate_version.clone());
+    } else {
+        log::error!("Unexpected dependency value type for {}", dep_key_str);
+        return;
     }
+
+    log::debug!("Setting {} to {}", dep_key_str, crate_version);
 }
```

Reply:
Understood. I initially thought to merge the functions and keep the original `update_dependencies` function and add check_dependency logic to it. But it kinda felt weird to both update and check dependencies by calling the same function which was either called `update_dependencies(...)` or `check_dependencies(...)`. It is not intuitive as to what the function does because it both updated and checked. That's why I decided to move with separate functions for each of the tasks.

Reply:
I am not saying this is definitely the 'right approach'.

@shawntabrizi what are your thoughts?